File: models/substance.py
from config import ORGANIC_TYPES, TOXIN_TYPES, INORGANIC_TYPES


class Substance:
    ORGANIC = 'ORGANIC'
    INORGANIC = 'INORGANIC'
    TOXIN = 'TOXIN'

    def __init__(
        self,
        name: str,
        type_: str,
        concentration: float,
        energy: float,
        volatility: float = 0.02
    ):
        self.name = name
        self.type = type_
        self.concentration = concentration
        self.energy = energy
        self.volatility = volatility  # 0.0 — стабильное, 0.9 — быстро распадается

    def update(self):
        """Естественное рассеивание."""
        self.concentration *= (1.0 - self.volatility)
        if self.concentration < 0.01:
            self.concentration = 0

    def is_active(self) -> bool:
        return self.concentration > 0

    @staticmethod
    def find_substance(name: str) -> tuple[str, dict] | None:
        found = None
        for group, type_list in [
            (Substance.ORGANIC, ORGANIC_TYPES),
            (Substance.TOXIN, TOXIN_TYPES),
            (Substance.INORGANIC, INORGANIC_TYPES),
        ]:
            for entry in type_list:
                if entry["name"] == name:
                    found = (group, entry)
                    break
            if found:
                break

        if not found:
            return None

        sub_type, data = found
        return sub_type, data

    def clone(self) -> 'Substance':
        """Создаёт копию вещества"""
        return Substance.from_dict(self.to_dict())

    def to_dict(self):
        return {
            "name": self.name,
            "type": self.type,
            "concentration": self.concentration,
            "energy": self.energy,
            "volatility": self.volatility,
        }

    @classmethod
    def from_dict(cls, data):
        return cls(
            name=data["name"],
            type_=data["type"],
            concentration=data["concentration"],
            energy=data["energy"],
            volatility=data["volatility"]
        )

    def __repr__(self):
        return (f"Substance(name={self.name}, type={self.type}, "
                f"conc={self.concentration:.3f}, energy={self.energy:.2f}, "
                f"volatility={self.volatility:.3f})")

File: java/evolution/src/main/java/com/life/evolution/model/Substance.java
package com.life.evolution.model;

import com.life.evolution.config.Config;

import java.util.HashMap;
import java.util.Map;

public class Substance {

    public static final String ORGANIC = "ORGANIC";
    public static final String INORGANIC = "INORGANIC";
    public static final String TOXIN = "TOXIN";

    private String name;
    private String type;
    private double concentration;
    private double energy;
    private double volatility;   // 0.0 stable, 0.9 fast decay

    public Substance(String name, String type, double concentration, double energy, double volatility) {
        this.name = name;
        this.type = type;
        this.concentration = concentration;
        this.energy = energy;
        this.volatility = volatility;
    }

    public Substance(String name, String type, double concentration, double energy) {
        this(name, type, concentration, energy, 0.01);
    }

    // ---------------------------------------------------------
    // UPDATE
    // ---------------------------------------------------------
    public void update() {
        concentration *= (1.0 - volatility);
        if (concentration < 0.01) {
            concentration = 0;
        }
    }

    public boolean isActive() {
        return concentration > 0;
    }

    // ---------------------------------------------------------
    // SUBSTANCE TYPE INFO (LOOKUP FROM CONFIG)
    // ---------------------------------------------------------
    public static Map<String, Object> findSubstance(String name) {
        return Config.SUBSTANCES.get(name);
    }

    // ---------------------------------------------------------
    // CLONE / COPY
    // ---------------------------------------------------------
    public Substance cloneSubstance() {
        return fromMap(toMap());
    }

    // ---------------------------------------------------------
    // JSON MAPPING
    // ---------------------------------------------------------
    public Map<String, Object> toMap() {
        Map<String, Object> m = new HashMap<>();
        m.put("name", name);
        m.put("type", type);
        m.put("concentration", concentration);
        m.put("energy", energy);
        m.put("volatility", volatility);
        return m;
    }

    @SuppressWarnings("unchecked")
    public static Substance fromMap(Map<String, Object> m) {
        return new Substance(
                (String) m.get("name"),
                (String) m.get("type"),
                ((Number) m.get("concentration")).doubleValue(),
                ((Number) m.get("energy")).doubleValue(),
                ((Number) m.getOrDefault("volatility", 0.01)).doubleValue()
        );
    }

    @Override
    public String toString() {
        return "Substance(" +
                "name=" + name +
                ", type=" + type +
                ", conc=" + String.format("%.3f", concentration) +
                ", energy=" + String.format("%.2f", energy) +
                ", volatility=" + String.format("%.3f", volatility) +
                ')';
    }

    // ---------------------------------------------------------
    // GETTERS / SETTERS
    // ---------------------------------------------------------
    public String getName() { return name; }
    public String getType() { return type; }
    public double getConcentration() { return concentration; }
    public void setConcentration(double c) { this.concentration = c; }
    public double getEnergy() { return energy; }
    public double getVolatility() { return volatility; }
}
